### DISCUSSION TITLE
Would it be possible for the Toolkit Nuget package to add the Community templates to VS

### DISCUSSION THREAD

**Discussion by yannduran** [VsixCommunity/Community.VisualStudio.Toolkit#17] [Ideas]
I've been wondering if it would be better for the Community templates to be installed by the Toolkit package instead of in a separate extension. The templates could either be 

- copied to the appropriate VS folder
- or copied to the user folder for templates (usually in the Documents, but even better would be to honor the user's settings)

![image](https://user-images.githubusercontent.com/2031126/115858864-d0b9da00-a472-11eb-9b1f-35d29082fbc2.png)

(sorry about the cutoff text, VS doesn't seem to be honoring my Windows 125% scale setting for text properly)

The benefits as I see it would be:

1. No extra extension would be needed to be installed by new extenders to get access to the Community templates.
2. **All** the code could be in the same place, in the same solution. It could then be edited/updated/deployed together (instead of an extension project **and** a Nuget package project).
3. One less moving part.
4. They would always be in sync together and could be versioned the same.

**Comment by madskristensen** [comment]
For many (most?) people, the use of the toolkit comes from creating a new project from the template pack. That means the template comes before the NuGet package. How would that work?

**Comment by yannduran** [comment]
Oh you're right! It's the old **_which comes first, the chicken or the egg_** problem, isn't it. Without a project there's nothing to install a Nuget package into. Damn!

In that case I'd like to suggest renaming the extension to `Community.VisualStudio.Templates` (Visual Studio Community Templates to match up with the Nuget package. 

What do you think of that idea?

**Reply by madskristensen** [reply]
Yeah, we can rename the repo that. What about the extension itself? Should it be renamed?

**Reply by yannduran** [reply]
Absolutely! (IMNSHO)

**Reply by reduckted** [reply]
> Oh you're right! It's the old **_which comes first, the chicken or the egg_** problem, isn't it. Without a project there's nothing to install a Nuget package into.

Also, a NuGet package shouldn't be putting files in other places in the file system. I'm not even sure NuGet packages have the ability to do that.

**Reply by yannduran** [reply]
Thanks, you could be right. But that was why I was asking if it was possible.

**Reply by CZEMacLeod** [reply]
@reduckted @yannduran The new SDK system actually does use NuGet packages to hold templates - and the `dotnet new` command works that way. VS2019 has an experimental option to use the new templates installed that way.
[Custom templates for dotnet new](https://docs.microsoft.com/en-us/dotnet/core/tools/custom-templates)
[Nuget.org Templates](https://www.nuget.org/packages?packagetype=template&sortby=relevance&q=&prerel=True)
[.NET CLI Templates in Visual Studio](https://devblogs.microsoft.com/dotnet/net-cli-templates-in-visual-studio/)
You can see an example at [MSBuild.SDK.SystemWeb.Templates](https://github.com/CZEMacLeod/MSBuild.SDK.SystemWeb/tree/main/src/MSBuild.SDK.SystemWeb.Templates). Although that does target an SDK style project too.
I don't know if there is any thoughts to the future of templates as Extensions vs. NuGet packages in the long term, and if there is likely to be a `Microsoft.NET.SDK.VSExtension` SDK type to replace the long form type currently used for VSIX - perhaps as part of the VS2022 timeline and the new 64bit extensions?
On a separate note - I did try using a [short form project](https://github.com/CZEMacLeod/Community.VisualStudio.Toolkit.VSGlobals.VSCT/blob/main/VSIXProject2/VSIXProject2.csproj), and it more or less worked - although some of the designers weren't happy - it did compile successfully.

**Comment by yannduran** [comment]
Thanks @CZEMacLeod for that info! That could be handy to know.

Unfortunately, we still come back to the **_chicken and the egg_** problem that Mads pointed out. 

Although, if a Community template is used, that's going to install the Nuget package, yes? So the Nuget package could then install the latest Community templates. So by using an existing Community template, the rest of the templates could be added/updated automatically.

Is my logic sound?

**Reply by CZEMacLeod** [reply]
Yes - the normal flow would be to install the templates (either as the VS extension from the gallery or marketplace, or the `dotnet new` template from nuget), and create a new project. 
The new project will have the reference to the runtime nuget package, and perhaps the VCST helper package.
Alternatively, a developer may start with an existing extension, or use the built in VS template, then manually install the nuget package(s) to make use of the functionality.
